backend-solidity/src/contracts/ryder-nft.sol

```diff
@@ -0,0 +1,223 @@
+    uint256 private tokenIdNonce = 5003;
+    uint256 private mintLimit = 2000;
+    string private nftTokenUriBase = "ipfs://ipfs/Qm../"; // without .json
+    string private nftTokenUriSuffix = ".json";
+    bool private metadataFrozen = false;
+
+    mapping(address => bool) minters;
+    mapping(address => bool) admins;
+
+    modifier adminOnly() {
+        require(admins[msg.sender], ERR_UNAUTHORIZED);
+        _;
+    }
+
+    constructor(address[] memory deployAdmins) {
+        supportedInterfaces[0x5b5e139f] = true; // ERC721Metadata
+        admins[msg.sender] = true;
+        for (uint256 i = 0; i < deployAdmins.length; ++i)
+            admins[deployAdmins[i]] = true;
+    }
+
+    // ERC721Metadata
+
+    function name() external pure returns (string memory _name) {
+        _name = NFT_NAME;
+    }
+
+    function symbol() external pure returns (string memory _symbol) {
+        _symbol = NFT_SYMBOL;
+    }
+
+    function tokenURI(uint256 tokenId)
+        external
+        view
+        validNFToken(tokenId)
+        returns (string memory)
+    {
+        return
+            string.concat(
+                nftTokenUriBase,
+                uint256ToString(tokenId),
+                nftTokenUriSuffix
+            );
+    }
+
+    function mint(address recipient) external {
+        require(minters[msg.sender], ERR_UNAUTHORIZED);
+        require(tokenIdNonce > mintLimit, ERR_ALREADY_DONE);
+        _mint(recipient, tokenIdNonce);
+        --tokenIdNonce;
+    }
+
+    function burn(uint256 tokenId) external canTransfer(tokenId) {
+        _burn(tokenId);
+    }
+
+    function getBalance(address who) external view returns (uint256) {
+        return _getOwnerNFTCount(who);
+    }
+
+    function tokenIdToTierId(uint256 _tokenId) public view returns (uint256) {
+        return
+            !dicksonParameterSet
+                ? 0
+                : (_tokenId**5 +
+                    dicksonParameter *
+                    _tokenId**3 +
+                    3002 *
+                    dicksonParameter**2 *
+                    _tokenId +
+                    dicksonParameter) % MAX_TOKENS;
+    }
+
+    function tier(uint256 tierId) public pure returns (uint256) {
+        if (tierId == 0) return 0;
+        if (tierId >= TIER_LOWER_BOUND_T2) return 1;
+        if (tierId >= TIER_LOWER_BOUND_T3) return 2;
+        if (tierId >= TIER_LOWER_BOUND_T4) return 3;
+        if (tierId >= TIER_LOWER_BOUND_T5) return 4;
+        if (tierId >= TIER_LOWER_BOUND_T6) return 5;
+        if (tierId >= TIER_LOWER_BOUND_T7) return 6;
+        return 7;
+    }
+
+    /**
+     *
+     */
+    function tierById(uint256 tokenId)
+        external
+        view
+        validNFToken(tokenId)
+        returns (uint256)
+    {
+        return tier(tokenIdToTierId(tokenId));
+    }
+
+    // Admin functions
+
+    function setDicksonParameter(uint256 _dicksonParameter) external adminOnly {
+        dicksonParameter = _dicksonParameter;
```

Comment:
This should be guarded by a check for `dicksonParameterSet`

Reply:
Should we use `time` here and do mod 5003 when setting the parameter?

Reply:
@friedger what is the upside to doing mod there?

Reply:
@MarvinJanssen It is easier to verify that we actually used the correct timestamp as promised.